File: 02Algorithm_Study/02_week_homeWork/bridge.cpp
#include <string>
#include <vector>
#include <queue>


using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    int answer = 0;
    int sum = 0;
    int i=0;
    queue<int> bridge;
    while(1){
    	if(i == truck_weights.size()){
    		answer += bridge_length;
    		break;
		}
    	answer++;
    	int tmp = truck_weights[i];
    	if(bridge.size() == bridge_length){
    		sum -= bridge.front();
    		bridge.pop();
    	}
    	
    	
    	if( tmp + sum <= weight){
    		bridge.push(tmp);
    		sum+=tmp;
    		i++;
    	}
    	
		else{
    		bridge.push(0);
    	}
    }
    return answer;
}

File: 02Algorithm_Study/02_week_homeWork/bridge.py
def solution(bridge_length, weight, truck_weights):
    answer = 0
    currentWeight = 0
    idx = 0
    bridge = []
    while(1):
        if(idx == len(truck_weights)):
            answer +=bridge_length
            break
        answer +=1
        tmp = truck_weights[idx]
        if(len(bridge) == bridge_length):
            currentWeight -=bridge[0]
            bridge.pop(0)
        if(tmp + currentWeight <= weight):
            bridge.append(tmp)
            currentWeight += tmp
            idx+=1
        else:
            bridge.append(0)

    return answer
